src/xunit.performance.api/Profilers/Etw/EtwListener.cs

```diff
@@ -0,0 +1,85 @@
+    {
+        public EtwListener(
+            EtwSessionData userSessionData,
+            IReadOnlyCollection<EtwUserProvider> userProviders,
+            IReadOnlyCollection<EtwKernelProvider> kernelProviders)
+        {
+            UserSessionData = userSessionData ?? throw new ArgumentNullException($"{nameof(userSessionData)} cannot be null.");
+            UserProviders = userProviders ?? throw new ArgumentNullException($"{nameof(userProviders)} cannot be null.");
+            KernelProviders = kernelProviders ?? throw new ArgumentNullException($"{nameof(kernelProviders)} cannot be null.");
+        }
+
+        public EtwSessionData UserSessionData { get; }
+
+        public IReadOnlyCollection<EtwUserProvider> UserProviders { get; }
+
+        public IReadOnlyCollection<EtwKernelProvider> KernelProviders { get; }
+
+        /// <summary>
+        /// Performs tasks associated with listening ETW events on Windows.
+        /// </summary>
+        /// <param name="runner">Operation to be profiled/traced.</param>
+        /// <returns>The return value of the method that the runner delegate encapsulates.</returns>
+        public TResult Record(Func<TResult> runner)
+        {
+            if (runner == null)
+                throw new ArgumentNullException($"{nameof(runner)} cannot be null.");
+
+            var fi = new FileInfo(UserSessionData.FileName);
+            var kernelFileName = Path.Combine($"{fi.DirectoryName}", $"{Path.GetFileNameWithoutExtension(fi.Name)}.kernel.etl");
+            var kernelProvider = KernelProviders.Aggregate(EtwKernelProvider.Default,
+                (current, item) => new EtwKernelProvider {
+                    Flags = current.Flags | item.Flags,
+                    StackCapture = current.StackCapture | item.StackCapture,
+                });
+            var needKernelSession = EtwHelper.NeedSeparateKernelSession(kernelProvider);
+
+            if (needKernelSession && !EtwHelper.CanEnableKernelProvider)
+            {
+                const string message = "The application is required to run as Administrator in order to capture kernel data.";
+                WriteErrorLine(message);
+                throw new InvalidOperationException(message);
+            }
+
+            TResult result;
+            WriteDebugLine("ETW capture start.");
+            using (var kernelSession = needKernelSession ? EtwHelper.MakeKernelSession(kernelFileName, UserSessionData.BufferSizeMB) : null)
+            {
+                kernelSession?.EnableKernelProvider(kernelProvider.Flags, kernelProvider.StackCapture);
+
+                using (var userSession = new EtwSession(UserSessionData))
+                {
+                    UserProviders.ForEach(provider => {
+                        userSession.EnableProvider(provider.Guid, provider.Level, provider.Keywords, provider.Options);
+                    });
+
+                    result = runner();
```

Comment:
You should wrap this in a try/catch or try/finally block to make sure that clean-up happens.

Reply:
In this case/function, the using blocks will cleanup the necessary objects. The only thing pending is the global state when using Pmc, and this happens before hands. I will handle it in a different PR.

---
In reply to: [143033645](https://github.com/Microsoft/xunit-performance/pull/241#discussion_r143033645) [](ancestors = 143033645)

Reply:
Ok, sounds good.

---
In reply to: [143236199](https://github.com/Microsoft/xunit-performance/pull/241#discussion_r143236199) [](ancestors = 143236199,143033645)